File: 백준/Silver/15654. N과 M （5）/N과 M （5）.java
import java.util.*;
import java.io.*;
public class Main {
	static int N, M;
	static int[] arr;
	static boolean[] visit;
	static List<Integer> chooses = new ArrayList<>();
	static StringBuilder sb = new StringBuilder();
	static void choose(int cur) {
		if (cur == M) {
			for (int i : chooses) {
				sb.append(i).append(" ");
			}
			sb.append("\n");
			return;
		}
		for (int i=0; i<N; i++) {
			if (visit[i]) continue;
			visit[i] = true;
			chooses.add(arr[i]);
			choose(cur + 1);
			chooses.remove(chooses.size() - 1);
			visit[i] = false;
		}
 	}
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		N = Integer.parseInt(st.nextToken());
		M = Integer.parseInt(st.nextToken());
		arr = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
		visit = new boolean[N];
		Arrays.sort(arr);
		choose(0);
		System.out.print(sb);
	}
}

File: 백준/Silver/15654. N과 M （5）/N과 M （5）.cc
#include <iostream>
#include <vector>
#include <algorithm>
#include <sstream>

using namespace std;

int N, M;
vector<int> arr;
vector<bool> visit;
vector<int> chooses;
stringstream ss;

void choose(int cur) {
    if (cur == M) {
        for (int i : chooses) {
            ss << i << " ";
        }
        ss << "\n";
        return;
    }
    for (int i = 0; i < N; i++) {
        if (visit[i]) continue;
        visit[i] = true;
        chooses.push_back(arr[i]);
        choose(cur + 1);
        chooses.pop_back();
        visit[i] = false;
    }
}


int main() {
    cin >> N >> M;
    arr.resize(N);
    visit.resize(N, false);

    for (int i = 0; i < N; i++) {
        cin >> arr[i];
    }

    sort(arr.begin(), arr.end());
    choose(0);

    cout << ss.str();
    return 0;
}
